File: CFD_Principles_and_Applications_Blazek/BookCode/C++/unstructured_2d/src/faceVectorsSymm.cpp
/// @file faceVectorsSymm.cpp
///
/// Correction of face vectors at symmetry boundaries.
///
//*****************************************************************************
//
//  (c) J. Blazek, CFD Consulting & Analysis, www.cfd-ca.de
//  Created February 15, 2014
//  Last modification: June 27, 2014
// 
//=============================================================================
//
//  This program is free software; you can redistribute it and/or
//  modify it under the terms of the GNU General Public License
//  as published by the Free Software Foundation; either version 2
//  of the License, or (at your option) any later version.
//
//  This program is distributed in the hope that it will be useful,
//  but WITHOUT ANY WARRANTY; without even the implied warranty of
//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
//  GNU General Public License for more details.
//
//  You should have received a copy of the GNU General Public License
//  along with this program; if not, write to the Free Software
//  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
//
//*****************************************************************************

#include "defs.h"
#include "geometry.h"

/// Corrects face vectors of edges which represent the symmetry boundary.
/// The reason is that there must be no component in the direction normal
/// to the boundary for the fluxes to be computed correctly. Function also
/// changes the boundary type to reflect the orientation of the symmetry
/// boundary: 501 for symmetry along x-direction, 502 along y-direction.
///
/// @exception std::bad_alloc  failed memory allocation
///
void Geometry::FaceVectorsSymm()
{
  int  i, j, ib, ibf, ibn, ie, ibegf, ibegn, iendf, iendn;
  int  *marker;
  REAL sx, sy;

  //  mark nodes at symmetry boundaries

  marker = new int[nndInt];
  for (i=0; i<nndInt; i++)
  {
    marker[i] = -1;
  }

  ibegf = 0;
  ibegn = 0;
  for (ib=0; ib<nSegs; ib++)
  {
    iendf = ibound[ib].bfaceIndex;
    iendn = ibound[ib].bnodeIndex;
    if (btype[ib]>=500 && btype[ib]<600)
    {
      sx = 0.0;
      sy = 0.0;
      for (ibf=ibegf; ibf<=iendf; ibf++)
      {
        sx += sbf[ibf].x;
        sy += sbf[ibf].y;
      }
      if (ABS(sx) > ABS(sy))
        btype[ib] = 501;      // symmetry at x=const. plane
      else
        btype[ib] = 502;      // symmetry at y=const. plane
      for (ibn=ibegn; ibn<=iendn; ibn++)
      {
        marker[bnode[ibn].node] = btype[ib] - 500;  // store symmetry axis
      }
    }
    ibegf = iendf + 1;
    ibegn = iendn + 1;
  }

  // correct face vectors where necessary

  for (ie=0; ie<nedInt; ie++)
  {
    i = edge[ie].i;
    j = edge[ie].j;
    if (marker[i]!=-1 && marker[j]!=-1)
    {
      if (marker[i] < 2)     // x=const. plane
        sij[ie].x = 0.0;
      else                   // y=const. plane
        sij[ie].y = 0.0;
    }
  }

  delete[] marker;
}

File: CFD_Principles_and_Applications_Blazek/BookCode/Fortran/unstructured_2d/src/faceVectorsSymm.f90
!> @file faceVectorsSymm.f90
!!
!! Correction of face vectors at symmetry boundaries.
!
! *****************************************************************************
!
!  (c) J. Blazek, CFD Consulting & Analysis, www.cfd-ca.de
!  Created February 25, 2014
!  Last modification: May 22, 2014
!
! *****************************************************************************
!
!  This program is free software; you can redistribute it and/or
!  modify it under the terms of the GNU General Public License
!  as published by the Free Software Foundation; either version 2
!  of the License, or (at your option) any later version.
!
!  This program is distributed in the hope that it will be useful,
!  but WITHOUT ANY WARRANTY; without even the implied warranty of
!  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
!  GNU General Public License for more details.
!
!  You should have received a copy of the GNU General Public License
!  along with this program; if not, write to the Free Software
!  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
!
! *****************************************************************************

!> Corrects face vectors of edges which represent the symmetry boundary.
!! The reason is that there must be no component in the direction normal
!! to the boundary for the fluxes to be computed correctly. Subroutine
!! also changes the boundary type to reflect the orientation of the symmetry
!! boundary: 501 for symmetry along x-direction, 502 along y-direction.
!!
!! @param marker  temporary memory for a node marker
!!
subroutine FaceVectorsSymm( marker )

  use ModDataTypes
  use ModGeometry
  use ModNumerics
  implicit none

! parameters
  integer :: marker(:)

! local variables
  integer     :: i, j, ib, ibf, ibn, ie, ibegf, ibegn, iendf, iendn
  real(rtype) :: sx, sy

! *****************************************************************************
! mark nodes at symmetry boundaries

  do i=1,nndint
    marker(i) = -1
  enddo

  ibegf = 1
  ibegn = 1
  do ib=1,nsegs
    iendf = ibound(1,ib)
    iendn = ibound(2,ib)
    if (btype(ib)>=500 .and. btype(ib)<600) then
      sx = 0.D0
      sy = 0.D0
      do ibf=ibegf,iendf
        sx = sx + sbf(1,ibf)
        sy = sy + sbf(2,ibf)
      enddo
      if (Abs(sx) > Abs(sy)) then
        btype(ib) = 501      ! symmetry at x=const. plane
      else
        btype(ib) = 502      ! symmetry at y=const. plane
      endif
      do ibn=ibegn,iendn
        marker(bnode(1,ibn)) = btype(ib) - 500    ! store symmetry axis
      enddo
    endif
    ibegf = iendf + 1
    ibegn = iendn + 1
  enddo

! correct face vectors

  do ie=1,nedint
    i = edge(1,ie)
    j = edge(2,ie)
    if (marker(i)/=-1 .and. marker(j)/=-1) then
      if (marker(i) < 2) then     ! x=const. plane
        sij(1,ie) = 0.D0
      else                        ! y=const. plane
        sij(2,ie) = 0.D0
      endif
    endif
  enddo

end subroutine FaceVectorsSymm
